homeassistant/components/ozw/config_flow.py

```diff
@@ -1,24 +1,175 @@
 """Config flow for ozw integration."""
+import logging
+
+import voluptuous as vol
+
 from homeassistant import config_entries
+from homeassistant.components.hassio.handler import HassioAPIError
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import AbortFlow
 
+from .const import CONF_INTEGRATION_CREATED_ADDON
 from .const import DOMAIN  # pylint:disable=unused-import
 
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ADDON_DEVICE = "device"
+CONF_ADDON_NETWORK_KEY = "network_key"
+CONF_NETWORK_KEY = "network_key"
+CONF_USB_PATH = "usb_path"
+CONF_USE_ADDON = "use_addon"
 TITLE = "OpenZWave"
 
+ON_SUPERVISOR_SCHEMA = vol.Schema({vol.Optional(CONF_USE_ADDON, default=False): bool})
+
 
 class DomainConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
     """Handle a config flow for ozw."""
 
     VERSION = 1
     CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
 
+    def __init__(self):
+        """Set up flow instance."""
+        self.addon_config = None
+        self.addon_info = None
+        self.network_key = None
+        self.usb_path = None
+        # If we install the add-on we should uninstall it on entry remove.
+        self.integration_created_addon = False
+
     async def async_step_user(self, user_input=None):
         """Handle the initial step."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
+
+        if not self.hass.components.hassio.is_hassio():
+            return self._async_use_mqtt_integration()
+
+        return await self.async_step_on_supervisor()
+
+    @callback
+    def _async_use_mqtt_integration(self):
+        """Handle logic when using the MQTT integration."""
         if "mqtt" not in self.hass.config.components:
             return self.async_abort(reason="mqtt_required")
+
+        return self.async_create_entry(
+            title=TITLE,
+            data={
+                CONF_USB_PATH: self.usb_path,
+                CONF_NETWORK_KEY: self.network_key,
+                CONF_INTEGRATION_CREATED_ADDON: self.integration_created_addon,
+            },
+        )
+
+    async def async_step_on_supervisor(self, user_input=None):
+        """Handle logic when on Supervisor host."""
         if user_input is not None:
-            return self.async_create_entry(title=TITLE, data={})
+            if user_input[CONF_USE_ADDON]:
+                return await self._async_use_addon()
+            return self._async_use_mqtt_integration()
+
+        return self.async_show_form(
+            step_id="on_supervisor", data_schema=ON_SUPERVISOR_SCHEMA
+        )
+
+    async def _async_use_addon(self):
+        """Handle logic when using the OpenZWave add-on."""
+        if await self._async_is_addon_running():
+            return self._async_use_mqtt_integration()
```

Comment:
We've discussed in the past that we eventually want the OZW integration to directly connect to the OZW add-on MQTT server so we are not relying on the MQTT integration to be set up. For easy migration, should we already store a boolean in the config entry if we get to this point in the code that we want to use the add-on. That way when we add direct connect support, we can bypass MQTT.

Reply:
So something like: `use_addon=True` or something.

Reply:
Yeah, for now let's store the inverse to the answer to the use add on as `use_mqtt_integration`. Later, when we need to longer require the MQTT integration, we can have a specific question for that.

Or should we not allow users to use the add-on in combination with the MQTT integration, ie only have one question?

Reply:
I'm reading our notes and we said that we should always use the connect MQTT connection for connecting to the add-on. So then we don't need an extra question.

Reply:
I'm storing the `use_addon` boolean option now in the config entry.